### PR TITLE
[WIP] Document LAPACK module

### PR DESCRIPTION
**Pull request by kshyatt**
Breakage due to docsystem horror/my own foolishness. cc @jakebolewski @yuyichao.

If people have comments on the working of the docstrings that's cool too. @andreasnoack, @dmbates, @jiahao.

### CODE REVIEW

File: doc/stdlib/linalg.rst [hunk 1]
```diff
@@ -917,7 +901,7 @@ Linear algebra functions in Julia are largely implemented by calling functions f
 
    Conjugate transpose array ``src`` and store the result in the preallocated array ``dest``\ , which should have a size corresponding to ``(size(src,2),size(src,1))``\ . No in-place transposition is supported and unexpected results will happen if ``src`` and ``dest`` have overlapping memory regions.
 
-.. function:: eigs(A, [B,]; nev=6, ncv=max(20,2*nev+1), which="LM", tol=0.0, maxiter=300, sigma=nothing, ritzvec=true, v0=zeros((0,))) -> (d,[v,],nconv,niter,nmult,resid)
```
Comment (tkelman) [inline doc/stdlib/linalg.rst]: this appears to be undoing a recent change

Comment (kshyatt) [inline doc/stdlib/linalg.rst]: Yeah I just realized. I rebased from an old version. Will fix.

**Comment by kshyatt** [comment]
Tests pass! Does anyone have comments on the actual docstring content?

File: base/linalg/lapack.jl [hunk 2]
```diff
@@ -107,6 +109,26 @@ for (gbtrf, gbtrs, elty) in
     end
 end
 
+"""
+    gbtrf!(kl, ku, m, AB)
+
+Compute the LU factorization of a banded matrix `AB`. `kl` is the first
+subdiagonal containing a nonzero band, `ku` is the last superdiagonal
+containing one, and `m` is the first dimension of the matrix `AB`. Returns
+the LU factorization in-place and `ipiv`, the vector of pivots used.
```
Comment (tkelman) [inline base/linalg/lapack.jl]: how are these returned, as `(AB, ipiv)` ?

Comment (kshyatt) [inline base/linalg/lapack.jl]: Exactly

File: base/linalg/lapack.jl [hunk 3]
```diff
@@ -157,16 +181,28 @@ for (gebal, gebak, elty, relty) in
     end
 end
 
+"""
+    gebal!(job, A)
+
+Balance the matrix `A` before computing its eigensystem or Schur factorization.
+`job` can be one of `N` (`A` will not be permuted or scaled), `P` (`A` will only
+be permuted), `S` (`A` will only be scaled), or `B` (`A` will be both permuted
+and scaled). Modifies `A` in-place and returns `ilo`, `ihi`, and `scale`. If
```
Comment (tkelman) [inline base/linalg/lapack.jl]: would also spell out the return tuple here

File: doc/stdlib/linalg.rst [hunk 4]
```diff
@@ -919,6 +919,8 @@ Linear algebra functions in Julia are largely implemented by calling functions f
 
 .. function:: eigs(A, [B,]; nev=6, ncv=max(20,2*nev+1), which="LM", tol=0.0, maxiter=300, sigma=nothing, ritzvec=true, v0=zeros((0,))) -> (d,[v,],nconv,niter,nmult,resid)
 
+.. function:: eigs(A, [B,]; nev=6, which="LM", tol=0.0, maxiter=300, sigma=nothing, ritzvec=true, v0=zeros((0,))) -> (d,[v,],nconv,niter,nmult,resid)
```
Comment (tkelman) [inline doc/stdlib/linalg.rst]: this should not be added

File: base/linalg/lapack.jl (hunk 2 above)
Comment (tkelman) [inline base/linalg/lapack.jl]: yeah i think that would help with clarity for multiple return fcns

File: base/linalg/lapack.jl [hunk 5]
```diff
@@ -107,6 +109,26 @@ for (gbtrf, gbtrs, elty) in
...
+subdiagonal containing a nonzero band, `ku` is the last superdiagonal
+containing one, and `m` is the first dimension of the matrix `AB`. Returns
+the LU factorization in-place and `ipiv`, the vector of pivots used.
+"""
+gbtrf!(kl::Integer, ku::Integer, m::Integer, AB::StridedMatrix)
+
+"""
+    gbtrs!(trans, kl, ku, m, AB, ipiv, B)
+
+Solve the equation `AB * X = B`. `trans` determines the orientation of `AB`.
```
Comment (tkelman) [inline base/linalg/lapack.jl]: what are the acceptable values for trans?

Comment (kshyatt) [inline base/linalg/lapack.jl]: I'll put that in :)

File: base/linalg/lapack.jl [hunk 6]
```diff
@@ -157,16 +181,28 @@ for (gebal, gebak, elty, relty) in
...
+`job` can be one of `N` (`A` will not be permuted or scaled), `P` (`A` will only
+be permuted), `S` (`A` will only be scaled), or `B` (`A` will be both permuted
+and scaled). Modifies `A` in-place and returns `ilo`, `ihi`, and `scale`. If
+permuting was turned on, `A[i,j] = 0` if `j > i` and `1 < j < ilo` or `j > ihi`.
+`scale` contains information about the scaling/permutations performed.
+"""
+gebal!(job::Char, A::StridedMatrix)
+
+"""
+    gebak!(job, side, ilo, ihi, scale, V)
```
Comment (tkelman) [inline base/linalg/lapack.jl]: describe what `side` does?

File: base/linalg/lapack.jl [hunk 7]
```diff
@@ -441,6 +486,91 @@ for (gebrd, gelqf, geqlf, geqrf, geqp3, geqrt, geqrt3, gerqf, getrf, elty, relty
     end
 end
 
+"""
+    gebrd!(A) -> (A, d, e, tauq, taup)
+
+Reduce `A` in-place to bidiagonal form `A = QBP'`. Returns `A`, containing the
+bidiagonal matrix `B`, `d`, containing the diagonal elements of `B`, `e`,
+containing the off-diagonal elements of `B`, `taup`, containing the
```
Comment (tkelman) [inline base/linalg/lapack.jl]: taup and tauq description order does not match return order?

File: base/linalg/lapack.jl [hunk 8]
```diff
@@ -441,6 +486,91 @@ for (gebrd, gelqf, geqlf, geqrf, geqp3, geqrt, geqrt3, gerqf, getrf, elty, relty
...
+elementary reflectors representing `P`, and `tauq` containing the
+elementary reflectors representing `Q`.
+"""
+gebrd!(A::StridedMatrix)
+
+"""
+    gelqf!(A, tau)
+
+Compute the `LQ` factorization of `A`, `A = LQ`. `tau` contains scalars
```
Comment (tkelman) [inline base/linalg/lapack.jl]: what dimensions does tau need to have?

File: base/linalg/lapack.jl [hunk 9]
```diff
@@ -441,6 +486,91 @@ for (gebrd, gelqf, geqlf, geqrf, geqp3, geqrt, geqrt3, gerqf, getrf, elty, relty
...
+    geqp3!(A, jpvt, tau)
+
+Compute the pivoted `QR` factorization of `A`, `AP = QR` using BLAS level 3.
+`P` is a pivoting matrix, represented by `jpvt`. `tau` stores the elementary
+reflectors. `A`, `jpvt`, and `tau` are modified in-place.
+"""
+geqp3!(A::StridedMatrix, jpvt::Vector{BlasInt}, tau::Vector)
+
+"""
+    geqrt!(A, T)
```
Comment (tkelman) [inline base/linalg/lapack.jl]: what dimensions does T need to have? and does this literally return `-> (A, T)` ?

File: base/linalg/lapack.jl [hunk 10]
```diff
@@ -536,6 +668,26 @@ for (tzrzf, ormrz, elty) in
...
+can be unmodified (`trans = N`), transposed (`trans = T`), or conjugate
+transposed (`trans = C`). Returns matrix `C` which is modified in-place
+with the result of the multiplication.
+"""
+ormrz!(side::Char, trans::Char, A::StridedMatrix, tau::StridedVector, C::StridedMatrix)
+
+"""
+    tzrzf!(A) -> (A, tau)
+
+Takes the upper trapezoidal matrix `A` to upper triangular form in-place.
```
Comment (tkelman) [inline base/linalg/lapack.jl]: would use a different word than "Takes" here - maybe Transforms or Reduces?

File: base/linalg/lapack.jl [hunk 11]
```diff
@@ -659,6 +811,50 @@ for (gels, gesv, getrs, getri, elty) in
...
 end
+
+"""
+    gels!(trans, A, B) -> (F, B, ssr)
+
+Solves the linear equation `A * X = B`, `A.' * X =B`, or `A' * X = B` using
+a QR or LQ factorization. Modifies the matrix/vector `B` in place with the
+solution. `A` is overwritten with its `QR` or `LQ` factorization. `trans`
+may be one of `N` (no modification), `T` (transpose), or `C` (conjugate
+transpose). `gels!` searches for the minumum norm/least squares solution.
```
Comment (tkelman) [inline base/linalg/lapack.jl]: minimum

File: base/linalg/lapack.jl [hunk 12]
```diff
@@ -775,7 +970,39 @@ for (gesvx, elty, relty) in
             X, rcond, ferr, berr, rpgf
         end
     end
- end
+end
+
+"""
+    gesvx!(fact, trans, A, AF, ipiv, equed, R, C, B) -> X, equed, R, C, B, rcond, ferr, berr, work
```
Comment (tkelman) [inline base/linalg/lapack.jl]: why no tuple parens here?

File: base/linalg/lapack.jl [hunk 13]
```diff
@@ -775,7 +970,39 @@ for (gesvx, elty, relty) in
...
+
+Solves the linear equation `A * X = B` (`trans = N`), `A.' * X =B`
+(`trans = T`), or `A' * X = B` (`trans = C`) using the `LU` factorization
+of `A`. `fact` may be `E`, in which case `A` will be equilibrated and copied
+to `AF`; `F`, in which case `AF` and `ipiv` from a previous factorization
+are inputs; or `N`, in which case `A` will be copied to `AF` and then
+factored. If `fact = F`, `equed` may be `N`, meaning `A` has not been
+equilibrated; `R`, meaning `A` was multiplied by `diagm(R)`; `C`, meaning `A`
```
Comment (tkelman) [inline base/linalg/lapack.jl]: there's a distinction between R and C of left-multiplying vs right-multiplying by the scaling matrices, right?

Comment (kshyatt) [inline base/linalg/lapack.jl]: I'm confused - what do you mean?

File: base/linalg/lapack.jl (hunk 8 above)
Comment (tkelman) [inline base/linalg/lapack.jl]: lost in rebase https://github.com/JuliaLang/julia/commit/bfb08ad03e552207cc0ebb8d0d7b416548b63c70#commitcomment-13407758

File: base/linalg/lapack.jl [hunk 14]
```diff
@@ -775,7 +993,41 @@ for (gesvx, elty, relty) in
...
+Solves the linear equation `A * X = B` (`trans = N`), `A.' * X =B`
+(`trans = T`), or `A' * X = B` (`trans = C`) using the `LU` factorization
+of `A`. `fact` may be `E`, in which case `A` will be equilibrated and copied
+to `AF`; `F`, in which case `AF` and `ipiv` from a previous `LU` factorization
+are inputs; or `N`, in which case `A` will be copied to `AF` and then
+factored. If `fact = F`, `equed` may be `N`, meaning `A` has not been
+equilibrated; `R`, meaning `A` was multiplied by `diagm(R)` from the left;
+`C`, meaning `A` was multiplied by `diagm(C)` from the right; or `B`, meaning
+`A` was multiplied by `diagm(R)` from the left and `diagm(C)` from the right.
+If `fact = F` and `equed = R,B` the elements of `R` must all be positive.
```
Comment (tkelman) [inline base/linalg/lapack.jl]: is this `equed` equals `R` or `B`, or literally `R,B` ?

Comment (jakebolewski) [inline base/linalg/lapack.jl]: 'N', 'R', 'C', or 'B'

Comment (tkelman) [inline base/linalg/lapack.jl]: Right, those are the allowable values for that input, but for this conditional sentence "the elements of `R` must all be positive," which values of `equed` does that apply to? I think an "or" would be clearer than a comma.

File: base/linalg/lapack.jl [hunk 15]
```diff
@@ -1014,6 +1262,14 @@ for (gglse, elty) in ((:dgglse_, :Float64),
     end
 end
 
+"""
+    gglse!(A, c, B, d) -> (X,res)
+
+Solves the equation `A * x = c` where `x` is subject to the equality
```
Comment (tkelman) [inline base/linalg/lapack.jl]: this is "solve" in a least squares sense, right? https://software.intel.com/en-us/node/521117 (though don't copy any wording directly from the mkl docs unless we check their doc license carefully)

File: base/linalg/lapack.jl [hunk 16]
```diff
@@ -1256,6 +1512,57 @@ for (geev, gesvd, gesdd, ggsvd, elty, relty) in
...
 end
+
+"""
+    geev!(jobvl, jobvr, A) -> (W, VL, VR)
+
+Finds the eigensystem of `A`. If `jobvl = N`, the left eigenvectors of
+`A` aren't computed. If `jobvr = N`, the right eigenvectors of `A`
+aren't computed. If `jobvl = V` or `jobvr = V`, the corresponding
+eigenvectors are computed. Returns the eigenvalues in `W`, the right
+eigenvectors in `VR`, and the left eigenvectors in `VL`.
```
Comment (tkelman) [inline base/linalg/lapack.jl]: sizes of W, VL, VR?

Comment (tkelman) [inline base/linalg/lapack.jl]: nevermind they're outputs not inputs

File: base/linalg/lapack.jl [hunk 17]
```diff
@@ -1256,6 +1512,57 @@ for (geev, gesvd, gesdd, ggsvd, elty, relty) in
...
+computed and returned separately. `jobu` and `jobvt` can't both be `O`.
+
+Returns `U`, `S`, and `Vt`, where `S` are the singular values of `A`.
+"""
+gesvd!(jobu::Char, jobvt::Char, A::StridedMatrix)
+
+"""
+    ggsvd!(jobu, jobv, jobq, A, B) -> (U, V, Q, alpha, beta, k, l, R)
+
+Finds the generalized singular value decomposition of `A` and `B`.
```
Comment (tkelman) [inline base/linalg/lapack.jl]: would be cool to spell out the matrix equation that this (and the normal svd's above) are solving

**Comment by kshyatt** [comment]
OK I've addressed the last N comments. If this passes CI I'm going to merge. If you want more changes either push commits yourself or open a PR.

**Comment by kshyatt** [comment]
The CI failure was in the OSX `backtrace` test.

**Comment by KristofferC** [comment]
Great job

**Comment by ViralBShah** [comment]
Should we backport this to 0.4?

**Comment by ScottPJones** [comment]
Is there a fix for the `backtrace` test, and what broke that?

**Comment by tkelman** [comment]
That was a separate commit on master. I have more editorial changes I'll be proposing in a PR but we could maybe backport the current version anyway.

**Pull request merged**